projects/extension/src/background/ConnectionManager.ts

```diff
@@ -194,25 +236,119 @@ export class ConnectionManager extends (EventEmitter as { new(): StateEmitter })
   async addChain(
     name: string,
     chainSpec: string,
-    jsonRpcCallback: SmoldotJsonRpcCallback): Promise<SmoldotChain> {
+    jsonRpcCallback: SmoldotJsonRpcCallback,
+    tabId?: number): Promise<Network> {
     if (!this.#client) {
       throw new Error('Smoldot client does not exist.');
     }
+    const existingNetwork = this.#networks.find(n => n.name === name && n.tabId === tabId)
+    if (existingNetwork)
+      return existingNetwork
 
     const addedChain = await this.#client.addChain({
       chainSpec,
       jsonRpcCallback,
       potentialRelayChains: this.#networks.map(net => net.chain),
     });
 
-    this.#networks.push({
+    const network: Network = {
+      tabId: tabId || 0,
       name,
       chain: addedChain,
       status: 'connected',
       isKnown: true,
       chainspecPath: `${name}.json`
-    });
+    }
+
+    this.#networks.push(network);
+
+    return network;
+  }
+
+  /** Handles the incoming message that contains Spec. */
+  #handleSpecMessage = (msg: MessageToManager, app: AppProps): void => {
+    const chainSpec: string = relayChains.has(app.chainName) ?
+      (relayChains.get(app.chainName) || '') : msg.payload;
+
+    const rpcCallback = (rpc: string) => {
+      const rpcResp = app.healthChecker?.responsePassThrough(rpc);
+      if (rpcResp)
+        app.port.postMessage({ type: 'rpc', payload: rpcResp })
+    }
+
+    this.addChain(app.chainName, chainSpec, rpcCallback, app.tabId)
+      .then(network => {
+        app.chain = network.chain;
+        // eslint-disable-next-line @typescript-eslint/unbound-method
+        app.healthChecker?.setSendJsonRpc(app.chain.sendJsonRpc);
+        app.healthChecker?.start((health: SmoldotHealth) => {
+          app.healthStatus = health;
+        });
+        // process any RPC requests that came in while waiting for `addChain`
+        // to complete
+        if (this.#pendingRequests.length > 0) {
```

Comment:
Also missed the test that messages sent before addChain resolves don't get lost. Again should be quick and easy to port ... fake posting some RPC messages before you `waitForMessageToBePosted` after the spec message

Reply:
Actually I have not - in tests now I am using the actual ConnectionManager and not the ConnectionManagerMock. I will try to figure out a way to test these, if possible with the ConnectionManager or else I will need to adjust the ConnectionManagerMock. Im not sure though if these should be tested with a mock - these are actually covered from our sample apps (that we use for test)